README.TXT
=head1 INTRODUCTION

Perl module for TrueType/OpenType font hacking. Supports reading, processing and
writing of the following tables: GDEF, GPOS, GSUB, LTSH, OS/2, PCLT,
bsln, cmap, cvt, fdsc, feat, fpgm, glyf, hdmx, head, hhea, hmtx, kern,
loca, maxp, mort, name, post, prep, prop, vhea, vmtx and the reading and
writing of all other table types.

In short, you can do almost anything with a standard TrueType font with
this module. Be Brave!

Any suggestions, improvements, additions, subclasses, etc. would be gratefully
received and probably included in a future release. Please send them to me.

This module has been tested on Win32, Unix and Mac.

Applications that were associated with this module have been moved to
Font::TTF::Scripts where great things can be done.

=head1 SYNOPSIS

Here is the regression test (you provide your own font). Run it once and then
again on the output of the first run. There should be no differences between
the outputs of the two runs.

    use Font::TTF::Font;

    $f = Font::TTF::Font->open($ARGV[0]);

    # force a read of all the tables
    $f->tables_do(sub { $_[0]->read; });

    # force read of all glyphs (use read_dat to use lots of memory!)
    # $f->{'loca'}->glyphs_do(sub { $_[0]->read; });
    $f->{'loca'}->glyphs_do(sub { $_[0]->read_dat; });
    # NB. no need to $g->update since $_[0]->{'glyf'}->out will do it for us

    $f->out($ARGV[1]);
    $f->DESTROY;               # forces close of $in and maybe memory reclaim!

=head1 INSTALLATION

If you have received this package as part of an Activestate PPM style .zip file
then type

    ppm install Font-TTF.ppd

Otherwise.

To configure this module, cd to the directory that contains this README file
and type the following.

    perl Makefile.PL

Alternatively, if you plan to install Font::TTF somewhere other than
your system's perl library directory. You can type something like this:

    perl Makefile.PL PREFIX=/home/me/perl INSTALLDIRS=perl

Then to build you run make.

    make

If you have write access to the perl library directories, you may then
install by typing:

    make install

To tidy up, type:

    make realclean

Windows users should use dmake instead of make.

=head1 CHANGES

See Changes for an overview of recent changes. 

=head2 Future Changes

I do not anticipate any more restructuring changes (but reserve the right to do so).

=head1 AUTHOR

Martin Hosken L<http://scripts.sil.org/FontUtils>.
(see CONTRIBUTORS for other authors).

=head1 LICENSING

Copyright (c) 1998-2016, SIL International (http://www.sil.org) 

This module is released under the terms of the Artistic License 2.0. 
For details, see the full text of the license in the file LICENSE.

The fonts in the test suite are released under the SIL Open Font License 1.1, see t/OFL.txt.

=cut


Changes
1.06  2016-08-17

* Source repo moved from Subversion to Github
* OpenType script and lang tags updated from ISO/IEC 14496-22:2015, draft amendment 2
* Various POD improvements
* Add deepcopy mode to Dumper::ttfdump()
* Bug fixes
*    Wasn't installing on Windows Perl 5.22 and up
*    Reading mark-to-ligature lookups would crash if anchors were omitted
*    Incorrect extension lookup structure
*    Multiple fixes in Silf table processing
*    rt.cpan.org #106562 Uninitialized value warnings
*    rt.cpan.org #106816 spelling errors in manpage

1.05  2015-01-26 r1069

* Add support for GSUB Type 8 Reverse-chaining substitution
* OpenType script/lang/feature tags now based on ISO/IEC 14496-22
* Remove deprecated GDL_old.pm
*
* Bug fixes:
*    Fix rt.cpan.org 92150, 93597
*	 Force 0xFFFF sentry to be in a segment by itself in format 4 cmap subtables
*    Less aggressive cmap format 4 optimization to eliminate USV holes
*    Fix various issues reading WOFF-compressed font tables
*    Fix reading DSIG

1.04  2014-01-09 r994

* Bug fixes:
*    Fix rt.cpan.org 80671, 80672; simplify fix for #42553 per OP
*    Shared tables in TTC weren't working
*    Quiet a cleanup warning in TTC processing
*    Update licensing info for test fonts to OFL

1.03  2013-11-10 r969

* Add $t->minsize() to all tables for assisting with font subsetting
* Add deduping to Name table writing
* Add OS/2 table method to guess at Unicode and Codepage range bits.
* Add support for cmap format 13 subtables
* Expunge notion of 'dflt' language tag from Ttopen.pm
* Bug fixes

1.02  2012-08-30 r862

* Fix typo in Useall to get case right

1.01  2012-08-30 r859

* Add IO::String prerequisite to make BSD and other testing environments happy

1.00  2012-08-28 r857

* Major change to glyph creation/editing -- now utilizes ' isDirty' flag
* Add support for:
*    V4 OS/2 table
*    Reading WOFF fonts
*    MarkFilterSets in GDEF
*    Feature parameters (for cvxx, ssxx and size features)
*    Additional Graphite tables (Silf, Glac, and Gloc)
* Updated MS Lang IDs to Dec 2011
* Finish implementation of coverage table sorting
* Copyright and licensing updated to CPAN requirements
* Bug fixes

0.48  2010-12-14 r692

* Ensure coverage tables are properly sorted
* Additional OT tags (from OpenType 1.6)
* Bug fixes

0.47  2009-08-10 r649

* Add support for Graphite Sill table
* Handle kerning tables version > 0
* Bug fixes

0.46  2009-01-26 r577

* Bug fixes
* Change Copyright to Artistic License 2.0

0.45  2008-06-11 r527

* Introduce changelog
* tidy up 0.44 package, fix README to be more accurate
* tests failing on perl 5.8.2 and before due to no use Exporter qw(import);
  Fix OTTags accordingly.

Some previous changes documented in lib/Font/TTF/Changes_old.txt


lib/Font/TTF.pm
package Font::TTF;

$VERSION = '1.05';   # MJPH    19-Jan-2015      Bug fixes; updated OT tags; GSUB Lookup Type 8 support
# $VERSION = '1.04';    # MJPH     8-Jan-2014      License, POD, and perl -w tidying; bug fixes
# $VERSION = '1.03';    # MJPH     5-Sep-2013      Add $t->minsize()
# $VERSION = '1.02';    # MJPH    30-Aug-2012      Fix case typo in Useall
# $VERSION = '1.01';    # MJPH    30-Aug-2012      add IO::String prerequisite
# $VERSION = '1.00';    # MJPH    21-Aug-2012      OS/2, OT & Graphite improvements; bug fixes
# $VERSION = '0.48';    # MJPH    15-DEC-2010      Bug fixes
# $VERSION = '0.47';    # MJPH     7-AUG-2009      Minor bug fix in Name.pm
# $VERSION = '0.46';    # MJPH    26-JAN-2009      Various bug fixes, add Sill table
# $VERSION = '0.45';    # MJPH    11-JUN-2008      Packaging tidying
# $VERSION = '0.44';    # MJPH     9-JUN-2008      Various bug fixes
# $VERSION = '0.43';    # MJPH    20-NOV-2007      Add a test!
# $VERSION = '0.42';    # MJPH    11-OCT-2007      Add Volt2ttf support
# $VERSION = '0.41';    # MJPH    27-MAR-2007      Remove warnings from font copy
#                                                  Bug fixes in Ttopen, GDEF
#                                                  Remove redundant head and maxp ->reads
# $VERSION = '0.40';    # MJPH    31-JUL-2006      Add EBDT, EBLC tables
# $VERSION = 0.39;

1;

=head1 NAME

Font::TTF - Perl module for TrueType Font hacking

=head1 DESCRIPTION

This module allows you to do almost anything to a TrueType/OpenType Font
including modify and inspect nearly all tables.

=head1 AUTHOR

Martin Hosken L<http://scripts.sil.org/FontUtils>.
(see CONTRIBUTORS for other authors).

=head1 LICENSING

Copyright (c) 1998-2016, SIL International (http://www.sil.org) 

This module is released under the terms of the Artistic License 2.0. 
For details, see the full text of the license in the file LICENSE.

.

=cut